File: TTdsa-cpp/Exercise/math/N12.cpp
//
// Created by renxin on 2021/6/18.
//

#include "ttdsa.h"

class Solution {
public:
    double Power(double base, int exponent) {
        if (exponent < 0) {
            return 1.0 / fun(base, -exponent);
        } else {
            return fun(base, exponent);
        }
    }

private:
    double fun(double base, int exponent) {
        if (exponent == 0)
            return 1.0;
        if (exponent == 1)
            return base;
        double part = fun(base, exponent / 2);
        if (exponent % 2 == 0)
            return part * part;
        else
            return part * part * base;
    }
};

int main() {
    Solution sol;
    // 8.00000
    cout << sol.Power(2.0, 3) << endl;
    // 9.26100
    cout << sol.Power(2.1, 3) << endl;
    // 0.25000
    cout << sol.Power(2.0, -2) << endl;
    return 0;
}

File: TTdsa-java/Exercise/src/com/aacirq/math/N12.java
package com.aacirq.math;

public class N12 {
    public static void main(String[] args) {
        N12 sol = new N12();

        // 8.00000
        System.out.println(sol.Power(2.0, 3));
        // 9.26100
        System.out.println(sol.Power(2.1, 3));
        // 0.25000
        System.out.println(sol.Power(2.0, -2));
    }

    public double Power(double base, int exponent) {
        if (base == 0)
            return 0;
        long e = exponent;
        double b = base;
        if (e < 0) {
            b = 1/b;
            e = -e;
        }
        double res = 1.0;
        while (e != 0) {
            if ((e & 1) != 0)
                res *= b;
            b *= b;
            e >>= 1;
        }
        return res;
    }
}
